state/estimator.go

```diff
@@ -0,0 +1,147 @@
+package state
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/celestiaorg/celestia-app/v3/pkg/user"
+	apptypes "github.com/celestiaorg/celestia-app/v3/x/blob/types"
+)
+
+// gasMultiplier is used to increase gas limit in case if tx has additional options.
+const gasMultiplier = 1.1
+
+var ErrGasPriceExceedsLimit = errors.New("state: estimated gasPrice exceeds max gasPrice")
+
+type estimator struct {
+	estimatorAddress string
+
+	defaultClientConn *grpc.ClientConn
+	estimatorConn     *grpc.ClientConn
+}
+
+func (e *estimator) Start(context.Context) error {
+	if e.estimatorAddress == "" {
+		return nil
+	}
+
+	conn, err := grpc.NewClient(
+		e.estimatorAddress,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithUnaryInterceptor(grpcRetry()),
+	)
+	if err != nil {
+		return err
+	}
+	e.estimatorConn = conn
+	return nil
+}
+
+func (e *estimator) Stop(context.Context) error {
+	if e.estimatorConn != nil {
+		if err := e.estimatorConn.Close(); err != nil {
+			return err
+		}
+	}
+	e.estimatorConn = nil
+	return nil
+}
+
+func (e *estimator) estimate(
+	ctx context.Context,
+	cfg *TxConfig,
+	client *user.TxClient,
+	msg sdktypes.Msg,
+) (float64, uint64, error) {
+	if cfg.GasPrice() != DefaultGasPrice && cfg.GasLimit() != 0 {
+		return cfg.GasPrice(), cfg.GasLimit(), nil
+	}
+
+	gasPrice, gasLimit, err := e.estimateGas(ctx, client, cfg.priority, msg)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if cfg.GasPrice() != DefaultGasPrice {
+		gasPrice = cfg.GasPrice()
+	}
+	if cfg.GasLimit() != 0 {
+		gasLimit = cfg.GasLimit()
+	}
+	if gasPrice > cfg.MaxGasPrice() {
+		return 0, 0, ErrGasPriceExceedsLimit
+	}
+	return gasPrice, gasLimit, nil
+}
+
+// estimateGas estimates gas in case it has not been set.
+func (e *estimator) estimateGas(
+	ctx context.Context,
+	client *user.TxClient,
+	priority TxPriority,
+	msg sdktypes.Msg,
+) (float64, uint64, error) {
+	signer := client.Signer()
+	rawTx, err := signer.CreateTx([]sdktypes.Msg{msg}, user.SetFee(1))
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to create raw tx: %w", err)
+	}
+
+	gasPrice, gas, err := e.queryGasUsedAndPrice(ctx, signer, priority, rawTx)
+	if err == nil {
+		return gasPrice, uint64(float64(gas) * gasMultiplier), nil
+	}
+
+	// Tx client will multiple estimated gas by 1.1 to cover additional costs
+	gas, err = client.EstimateGas(ctx, []sdktypes.Msg{msg})
+	if err != nil {
+		return 0, 0, fmt.Errorf("state: failed to estimate gas: %w", err)
+	}
+	return DefaultGasPrice, gas, nil
+}
+
+func (e *estimator) queryGasUsedAndPrice(
+	ctx context.Context,
+	signer *user.Signer,
+	priority TxPriority,
+	rawTx []byte,
+) (gasPrice float64, gas uint64, err error) {
+	conns := []*grpc.ClientConn{e.estimatorConn, e.defaultClientConn}
```

Comment:
So the idea here is that there will /can be an estimator service that will perform this function, but if not, the default validator you're using to submit txs will also be able to estimate.

Reply:
From what I understand your comment is just informational, nothing to change, right?